16.0 mig payment slimpay issue

fcayre opened the pull request.

fcayre committed:
Import the payment_slimpay_issue module into the 12.0 branch

legalsylvain committed:
[ADD] migration analysis files

fcayre committed:
Display project task partners on kanban view

fcayre committed:
Fix payment_slimpay_issue tests in the CI

The reason for the crash was that in the CI, the commown module is
installed, overriding the project.task's
slimpay_payment_issue_process_automatically method. We want to test
current module's implementation and do not want to rewrite the tests
with contract-generated invoices (required by the commown module's
implementation), as it is slow and cumbersome.

The chosen solution was so to monkey patch the commown's implementation
when the commown module is installed (i.e. importable) and in the
present tests only.

Another error had the same source: commown module being installed in
the CI context. In this case, a SMS was sent in the scenario of
test_functional_3_trials, which was not the case out of the CI and
defeated the tests related to the sent emails, which had to be fixed.

legalsylvain committed:
[META] enable flake8

legalsylvain committed:
[META] enable black

legalsylvain committed:
[META] enable pyupgrade

legalsylvain committed:
[META] enable isort

legalsylvain committed:
[META] enable black (new version)

fcayre committed:
Fix double payment of contract-related issues

s2s_do_transaction recreates a payment although created just before
being called in this module. This is because the server action in
product_rental if trigger by the transaction state update and will
create a payment when the transaction is related to a contract (which
is not the case in present module's unit-tests but can easily be
simulated by modifying the server action's code).

Commit:
Modifie payment vue so payment tab appears only in relevant project

Commit:
i18n + l10n-fr

fcayre committed:
payment_slimpay_issue 12.0.1.0.1

fcayre committed:
Move test code from payment_slimpay_issue to commown_payment_slimpay_issue

This test code was protecting against reverse dependency modules
installation, typically in the CI, which breaks the
payment_slimpay_issue module's tests.

This is now in the right order: the modules that modify the expected
test behaviour of their dependency are in charge of keeping keeping
their tests green.

Of course, an even better behaviour would be that the CI do not install
all modules before launching their tests but launch them in their own
installed environment (like we do locally).

fcayre committed:
Fix logging error and flake8 conformance

github-grap-bot committed:
payment_slimpay_issue 12.0.1.0.2

fcayre committed:
Coding style improvements

fcayre committed:
No more ignore some invoices in the semi-automated payment issue handling

Remove slimpay_payment_issue_process_automatically override that used to return
true only for contract-issued invoices.
This was a problem for merged invoices, which were to be handled manually, but
we choose here to handle automatically all cases Slimpay incidents for which an
invoice could be associated with: there is likely no situation, if any, where
this would cause a problem.

fcayre committed:
Handle payment issues automatically when related invoice payment method is electronic in

github-grap-bot committed:
[BOT] post-merge updates

Commit:
Merge commown_payment_slimpay_issue into payment_slimpay_issue

Abandon the idea of pushing payment_slimpay_issue into OCA.
Therefore simplify the code.

Commit:
i18n fr10n

github-grap-bot committed:
[BOT] post-merge updates

fcayre committed:
[tests] Mute some loggers in tests that check errors

Commit:
Render the template before converting the html

Commit:
Add partner language in context when rendering sms template

github-grap-bot committed:
[BOT] post-merge updates

fcayre committed:
[tests] Fix spurious warnings in tests

... either by muting some loggers or by changing the test code slightly

fcayre committed:
Turn a log's level from warning to info as it is a normal situation

github-grap-bot committed:
[BOT] post-merge updates

fcayre committed:
[fr t10n] Remove duplicate message

github-grap-bot committed:
[BOT] post-merge updates

fcayre committed:
Add missing i18n pot files

fcayre committed:
Post reject reason message in Slimpay issue tasks

fcayre committed:
i18n + partial fr t10n update

github-grap-bot committed:
[BOT] post-merge updates

fcayre committed:
Add readable Slimpay reject error text

fcayre committed:
i18n update + complete FR t10n

github-grap-bot committed:
[BOT] post-merge updates

fcayre committed:
Turn an error log into a proper exception to see the message in queue jobs

fcayre committed:
Retry payment using last posted payment partner's payment token, if any

Else try invoice partner's payment_token_id. This should never be the
case in real life however, as if you retry the payment... a previous one
exists! So this mainly eases tests.

Depend on contract_payment_auto to benefit from the payment_token_id
field of partners.

fcayre committed:
Add an error test to improve coverage

github-grap-bot committed:
[BOT] post-merge updates

fcayre committed:
[IMP] payment_slimpay_issue: black, isort, prettier

fcayre committed:
[MIG] payment_slimpay_issue: Migration to 16.0

fcayre committed:
[tests] Account user_type_id field was turned into the Selection typed account_type

fcayre committed:
[tests] env.user.company_id is now directly accessible at env.company

fcayre committed:
[tests] Change SavepointCase into TransactionCase and use @tagged

fcayre committed:
[tests] Adapt to the new file structure of the account_payment_slimpay module

fcayre committed:
[tests] Fix account codes (cannot contain '_')

fcayre committed:
[tests] Use xml id to get the slimpay provider

fcayre committed:
[tests] Simplify super() call

fcayre committed:
Invoice is now account.move and has several attributes renames

fcayre committed:
Payment acquirers have become payment providers and slimpay_client is now a callable

fcayre committed:
Electronic inbound payment instance has disappeared from the payment module

So instead of hard-coding this one, use its payment_type and code attributes
to check for electronic 'in' payment method usage.

fcayre committed:
[tests] Replace unittests mocks by requests mocks

This improves test coverage and descreases complexity.
Moreover, some model method mocks simply do not work anymore as these methods
have become read-only.

fcayre committed:
Adapt html mail template to the new format

fcayre committed:
SMS templates are now sms.template, no more mail.template

fcayre committed:
Payment token name was replaced by payment_details

fcayre committed:
Use register payment wizard instead of manually instanciating a transaction and a payment

This is much simpler and does all the hard work:
- create the payment
- create the transaction
- send the payment request to the provider

fcayre committed:
Fix reconciliation removal: payment_move_line_ids do not exist anymore

fcayre committed:
Invoices changed attribute and action names

fcayre committed:
Invoice payments are now accessed through reconcile partials

fcayre committed:
Execute tracking instructions after slimpay issue task creation

This aims to allow the attribute tracking actions that follow to take
place in the same transaction.

The task creation inhibits further attribute tracking actions for the
whole transaction by calling mail_thread._track_discard.

This is not appropriate here as we want an email to be sent when the
task is moved to a stage with a mail_template_id, which is done through
the tracking mechanism in Odoo 16.0.

fcayre committed:
[tests] Journal update_posted attribute was removed

Also improve code readability a bit.

fcayre committed:
[tests] Remove customer account definition as setting it on partner is not necessary

... and this simplifies the code somewhat.

fcayre committed:
[tests] Create an electronic payment method as there is none in our dependency modules anymore

fcayre committed:
[tests] Adapt to new payment.token attribute names

fcayre committed:
[tests] Adapt invoice creation to its new data model (account.move)

fcayre committed:
[tests] Replace transaction+payment creation by account.payment.register wizard

This comes with the cost of setting-up a requests mocker but it is simpler
to understand and less code to maintain.

fcayre committed:
[tests] We now have a direct access to transactions from the invoice: use it

... and suppress a now useless method, to simplify things.

fcayre committed:
[tests] Adapt invoice and payment state checks

Invoices only have 2 states: draft and posted. This is not interesting for
us, so we rather use invoice's payment_state, which is computed from the
invoice's payments.

We once test the payment state so we store it as an attribute of the test.

fcayre committed:
[tests] payment acquirers were renamed to payment providers

fcayre committed:
[tests] Make Slimpay payment references unique

This avoids a crash in the code were we retrieve several
payment.transaction instances instead of a single one when they
have the same Slimpay reference while this is not supposed to
occur.

fcayre committed:
[tests] Rename all payment references to better match Slimpay conventions

Ths new name looks like one Slimpay would have used, making it
easier to understand the tests.

fcayre committed:
[tests] Adapt invoice attribute names

fcayre committed:
[tests] Fix reconciliation removal: payment_move_line_ids do not exist anymore

fcayre committed:
[tests] Odoo exception name attribute is now deprecated

fcayre committed:
[tests] SMS template xml id, post method and queue job function name changes

fcayre committed:
[tests] Set explicit product taxes to zero to avoid a price check failure

fcayre committed:
[tests] Replace payment token unlink (which crashes) by removal from its partner

Unlink crashes because existing payment transaction relation to the
token is mandatory.

fcayre committed:
[tests] Add test pre-requisite checks

fcayre committed:
[tests] Explicitely flush and run precommit ops after task creation and task stage update

This mimics real-life transactions split:
- one for the task creation
- one for the stage update

This is required because we perform the task creation and then the stage
update in the same transaction, whereas the attribute tracking mechanism,
responsible for sending the email when the task stage is modified, is
inhibited for the entire db transaction after a creation.

fcayre committed:
[tests] Handle multi-company in the tests

Honeyxilia commented on payment_slimpay_issue/models/project_task.py:
```diff
@@ -417,7 +417,7 @@ def _slimpay_payment_issue_handle(self, project, client, issue_doc):
 
         _logger.info('Unreconciling invoice "%s"', invoice.name)
-        invoice.payment_move_line_ids.remove_move_reconcile()
+        invoice.line_ids.remove_move_reconcile()
```
Plus une remarque sur le message de commit, mais le changement semble arriver plus sur l'attribut que la méthode - ainsi, je pense qu'il serait plus correct de dire que l'attribut a changé

Honeyxilia commented on payment_slimpay_issue/models/project_task.py:
```diff
@@ -236,7 +236,7 @@ def _slimpay_payment_issue_get_or_create(self, project, client, issue_doc):
             "Slimpay Id: %s" % issue_doc["id"],
```
Quelques petites fautes dans le message de commit c:
- "allow *coming-up* tracking actions" -> "allow *follow-up* tracking actions" (si c'est bien pour parler d'actions qui suvient, comme le mail)
- "mecanism" -> "mechanism"

Honeyxilia commented on payment_slimpay_issue/tests/test_project_task.py:
```diff
@@ -482,7 +482,7 @@ def test_actions(self):
 
         # Prepare to new payment:
-        self.invoice.payment_move_line_ids.remove_move_reconcile()
+        self.invoice.line_ids.remove_move_reconcile()
```
Même remarque qu'avant, concernant le changement d'attribut

Honeyxilia commented on payment_slimpay_issue/tests/test_project_task.py:
```diff
@@ -815,9 +815,10 @@ def test_db_savepoint(self):
 
```
fixup à fold c:

Honeyxilia commented on payment_slimpay_issue/tests/test_project_task.py:
```diff
@@ -471,12 +471,19 @@ def _simulate_wait(self, task, check_job_function=False, **timedelta_kwargs):
             trap.assert_jobs_count(1, only=check_job_function)
```
pour le message du commit : "mecanism" -> "mechanism"

Honeyxilia commented on payment_slimpay_issue/tests/test_project_task.py:
```diff
@@ -56,11 +56,26 @@ def fake_issue_doc(
 
```
un commit nommé WIP : Est-ce qu'il doit être renommé ?

fcayre replied:
Oui tout à fait.

Honeyxilia reviewed (commented):
Quelques commits à modifier (surtout le fixup et le WIP) mais après ça, ça me paraît bien c:

fcayre committed:
Fix and test payment fees

We now use raise_if_not_found=False parameter of env.ref to return None when
the concerned payment fees product has been deleted (which can be a way to
handle them manually).

Before the patch, this was done correctly only for the bank (supplier) fees.
This is now also the case for our management fees. Both now use the same
method.

fcayre committed:
[tests] Remove now obsolete comment on payment reference

fcayre committed:
[tests] Factor out a Slimpay payment context manager and use it

This reduces test code duplication and will make it easier to add tests.

fcayre committed:
[tests] Improve coverage: test without bank fees product

fcayre committed:
Remove code that silences a Slimpay authentication error

This code was targeted at dev environments where Slimpay authentication
issues generate useless logs.

However the crontab is deactivated in a dev environment so there is no
spurious log generated and we can remove this (untested) code.

fcayre committed:
[tests] Improve coverage: check all Slimpay issue pages are fetched

fcayre committed:
Fix wrong Slimpay parameter used to detect the reject reason code

This was already a bug in v12 (see ticket ID 24174).

fcayre committed:
Better handling of Slimpay rejects with an unknown reason

Slimpay's advertised reason is now always printed.
A specific test was added.

fcayre committed:
[tests] Add a partner to a Slimpay payment issue doc to improve test coverage

We now test the code that attributes a partner to a task issue even
when no corresponding invoice was found.

fcayre committed:
[tests] Remove defensive test code line from coverage measurement

fcayre committed:
[tests] Robustify test to avoid negative cached sorted effects on invoice transactions order

There can be a difficult to debug order change if result set 'sorted' method
is used somewhere else and changes the default ('id desc') transaction order.

To avoid any problem, explicitely sort the way the test needs it.

fcayre committed:
Fix and test payment token choice for payment issue retry

In 16.0, invoice payment are no longer easily accessible once
unreconciled, so is last payment token, which is the preferred
token for payment retry.
However the last invoice transaction can play the exact same
role and is much easier to find.

fcayre committed:
[tests] Add a test to improve coverage: Slimpay partner ref not found

Honeyxilia commented on payment_slimpay_issue/tests/test_project_task.py:
```diff
@@ -368,7 +368,7 @@ def test_cron_first_issue(self):
 
```
erreur dans le message de commit : 
- We *know* test -> We *now* test

Honeyxilia commented on payment_slimpay_issue/models/project_task.py:
```diff
@@ -356,9 +356,9 @@ def _slimpay_payment_issue_retry_payment(self):
             invoice = task.invoice_id
```
Erreur dans message de commit : 
- are no *more* easily -> are no *longer* easily

Honeyxilia reviewed (approved):
A part des fautes, ça me parait ok c:

fcayre commented:
/ocabot merge patch

fcayre commented:
/ocabot rebase

fcayre commented:
/ocabot merge patch

fcayre commented:
/ocabot merge patch
